src/field_theory/complete_ordered.lean

```diff
@@ -0,0 +1,769 @@
+-/
+
+noncomputable theory
+open_locale classical
+
+open set
+
+/-- The natural equiv between the rationals and the rationals as a set inside any characteristic 0
+    division ring -/
+def rat_equiv_of_char_zero (F : Type*) [division_ring F] [char_zero F] : ℚ ≃ range (coe : ℚ → F) :=
+equiv.set.range coe rat.cast_injective
+
+/-- The natural equiv between the rationals as a set inside any pair of characteristic 0
+    division rings -/
+def range_rat_equiv (F K : Type*) [division_ring F] [char_zero F] [division_ring K] [char_zero K] :
+  range (coe : ℚ → F) ≃ range (coe : ℚ → K) :=
+(rat_equiv_of_char_zero F).symm.trans (rat_equiv_of_char_zero K)
+
+@[simp]
+lemma range_rat_equiv_apply {F K : Type*} {q : ℚ}
+  [division_ring F] [char_zero F] [division_ring K] [char_zero K] :
+  range_rat_equiv F K ⟨q, mem_range_self q⟩ = ⟨q, mem_range_self q⟩ :=
+by simp only [range_rat_equiv, rat_equiv_of_char_zero, equiv.set.range_apply,
+  function.comp_app, subtype.mk_eq_mk, equiv.coe_trans, rat.cast_inj, equiv.symm_apply_eq]
+
+/-- The function sending subsets of the rationals embedded inside of one characteristic zero
+    division ring to the corresponding subset of a second such ring. -/
+def cut_map (F K : Type*) [division_ring F] [char_zero F] [division_ring K] [char_zero K] :
+  set (set.range (coe : ℚ → F)) → set (set.range (coe : ℚ → K)) :=
+λ c, (range_rat_equiv F K) '' c
+
+/- No longer used -/
+lemma bdd_above_add {F : Type*} [ordered_add_comm_monoid F] {A B : set F} :
+  bdd_above A → bdd_above B → bdd_above (A + B) :=
+begin
+  rintros ⟨bA, hbA⟩ ⟨bB, hbB⟩,
+  use bA + bB,
+  rintros x ⟨xa, xb, hxa, hxb, rfl⟩,
+  exact add_le_add (hbA hxa) (hbB hxb),
+end
+
+
+/-- Equivalence commuting with multiplicative, additive and order structure. -/
+@[nolint has_inhabited_instance]
+structure ordered_ring_equiv (R S : Type*) [has_mul R] [has_add R] [has_mul S] [has_add S]
+  [has_le R] [has_le S] extends R ≃+* S :=
+(map_rel_iff' : ∀ {x y : R}, to_fun x ≤ to_fun y ↔ x ≤ y)
+
+/-- Reinterpret an ordered ring equivalence as a ring equivalence. -/
+add_decl_doc ordered_ring_equiv.to_ring_equiv
+
+-- /-- Reinterpret an ordered ring equivalence as an equivalence of types. -/
+-- add_decl_doc ordered_ring_equiv.to_equiv
+
+namespace ordered_ring_equiv
+
+infix ` ≃+*o `:25 := ordered_ring_equiv
+
+variables {R S : Type*} [has_mul R] [has_add R] [has_le R] [has_mul S] [has_add S] [has_le S]
+
+/-- Reinterpret an ordered ring equivalence as an order isomorphism. -/
+def to_order_iso (f : R ≃+*o S) : R ≃o S := { ..f }
+
+instance has_coe_to_ring_equiv : has_coe (R ≃+*o S) (R ≃+* S) := ⟨ordered_ring_equiv.to_ring_equiv⟩
+instance has_coe_to_mul_equiv : has_coe (R ≃+*o S) (R ≃* S) :=
+⟨λ f, (ordered_ring_equiv.to_ring_equiv f).to_mul_equiv⟩
+instance has_coe_to_add_equiv : has_coe (R ≃+*o S) (R ≃+ S) :=
+⟨λ f, (ordered_ring_equiv.to_ring_equiv f).to_add_equiv⟩
+instance has_coe_to_order_iso : has_coe (R ≃+*o S) (R ≃o S) := ⟨to_order_iso⟩
```

Comment:
Is this necessary vs just implementing one layer of casting and letting it apply transitively?

Reply:
Probably I went a bit over the top here, removed the unneeded ones thanks!